readme.md
# Scaryshot [![Build Status](https://travis-ci.org/applieddataconsultants/scaryshot.svg?branch=master)](https://travis-ci.org/applieddataconsultants/scaryshot)

Phantom.js screenshot web service

![scaryshot](https://raw.github.com/applieddataconsultants/scaryshot/master/scaryshot.png)

## Setup

Requires that you have the [phantom.js](http://phantomjs.org/) binary in your PATH.

## Installation

[![NPM](https://nodei.co/npm/scaryshot.png?downloads=true)](https://nodei.co/npm/scaryshot)

Simple install is:

```sh
(sudo) npm install -g scaryshot
scaryshot [port]
```

Then visit the webpage in your browser for usage.

## Currently supported features

* Formats (PDF, PNG, JPEG, GIF)
* Headers and Footers (PDF only)
* Delayed Snapshots
* Raw HTML or a URL as input
* Custom output filename

# License

The MIT License (MIT)

Copyright (c) 2014 Applied Data Consultants

Permission is hereby granted, free of charge, to any person obtaining a copy of
this software and associated documentation files (the "Software"), to deal in
the Software without restriction, including without limitation the rights to
use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
the Software, and to permit persons to whom the Software is furnished to do so,
subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


changelog.md
1.0.0 / 2016-06-07
==================

-	Remove margin/border on pdf
-	Switch to tape for tests

0.2.0 / 2014-06-12
==================

-	add; orientation support

0.1.1 / 2014-05-30
==================

-	add; max timeout kill to phantomjs process
-	add; preferGlobal flag
-	fix; node 0.11 stream issue

0.1.0 / 2013-10-08
==================

-	add; test suite (travis integration)
-	add; cleanup tmp files after generate

0.0.6 / 2013-10-08
==================

-	fix; 'argument list too long' for larger inputs

0.0.5 / 2013-10-08
==================

-	fix; envvars defaulting to string "undefined"

0.0.4 / 2013-10-08
==================

-	add; support for pdf headers and footers

0.0.3 / 2013-08-05
==================

-	fixed; os.tmpdir may not always have / at end
-	added; logo
-	added; license text
-	added; forkme

0.0.2 / 2013-08-01
==================

-	updated; better approach to raw html content
-	added; raw html renders, POST method

0.0.1 / 2013-07-31
==================

-	fixed; absolute path to render


test/scaryshot-test.js
var request = require('supertest')
var test = require('tape')
var fs = require('fs')
var app = require('../')
var sampleUrlWithProtocol = 'http://www.google.com/'
var sampleUrl = 'www.google.com'

function end(t) {
  return function(er) {
    t.notOk(er)
    t.end()
  }
}

test('/generate', function(ts) {
  ts.test('should fail if no url or html param supplied', function(t) {
    request(app)
      .get('/generate')
      .expect('content-type', 'application/json')
      .end(function(er, res) {
        t.equal(res.body.status, 400, 'expected 400 status property')
        t.ok(res.body.message, 'expected an error message')
        t.end()
      })
  })

  ts.test('should fail if invalid type param supplied', function(t) {
    request(app)
      .post('/generate')
      .type('form')
      .send({url: 'google.com', type: 'tiff'})
      .end(function(er, res) {
        t.equal(res.body.status, 400, 'expected 400 status property')
        t.ok(/tiff/.test(res.body.message), 'expected tiff within error message')
        t.end()
      })
  })

  ts.test('should accept url with protocol', function(t) {
    request(app)
      .get('/generate')
      .query({url: sampleUrlWithProtocol})
      .expect('content-type', 'application/pdf')
      .expect(200, end(t))
  })

  ts.test('should accept url without protocol', function(t) {
    request(app)
      .post('/generate')
      .type('form')
      .send({url: sampleUrl})
      .expect('content-type', 'application/pdf')
      .expect(200, end(t))
  })
  ts.test('should generate png files', function(t) {
    request(app)
      .get('/generate')
      .query({url: sampleUrl, type: 'png'})
      .expect('content-type', 'image/png')
      .expect(200, end(t))
  })
  ts.test('should generate gif files', function(t) {
    request(app)
      .post('/generate')
      .type('form')
      .send({url: sampleUrl, type: 'gif'})
      .expect('content-type', 'image/gif')
      .expect(200, end(t))
  })
  ts.test('should generate jpeg files', function(t) {
    request(app)
      .get('/generate')
      .query({url: sampleUrl, type: 'jpeg'})
      .expect('content-type', 'image/jpeg')
      .expect(200, end(t))
  })
  ts.test('should generate jpg files', function(t) {
    request(app)
      .post('/generate')
      .type('form')
      .send({url: sampleUrl, type: 'jpg'})
      .expect('content-type', 'image/jpeg')
      .expect(200, end(t))
  })
  ts.test('should generate output from html', function(t) {
    request(app)
      .get('/generate')
      .query({html: '<h1>Hello World</h1>', type: 'png'})
      .expect('content-type', 'image/png')
      .expect(200, end(t))
  })
  ts.test('should handle large html input', function(t) {
    request(app)
      .post('/generate')
      .type('form')
      .send({html: fs.readFileSync(__dirname + '/fixtures/large.html').toString(), type: 'png'})
      .expect('content-type', 'image/png')
      .expect(200, end(t))
  })
})


render.js
var page = require('webpage').create()
var system = require('system')

var env = JSON.parse(system.stdin.read())

var type = env.PHANTOM_TYPE || 'pdf'
var header = env.PHANTOM_HEADER || null
var footer = env.PHANTOM_FOOTER || null
var orientation = env.PHANTOM_ORIENTATION || 'portrait'

page.settings.localToRemoteUrlAccessEnabled = true

if (type == 'pdf')
  page.paperSize = {
    format: 'Letter',
    orientation: orientation,
    header: header && {
      height: '0.9cm',
      contents: phantom.callback(function(pageNum, numPages) {
        return header.replace(/{{page}}/g, pageNum).replace(/{{pages}}/g, numPages)
      }),
    },
    footer: footer && {
      height: '0.9cm',
      contents: phantom.callback(function(pageNum, numPages) {
        return footer.replace(/{{page}}/g, pageNum).replace(/{{pages}}/g, numPages)
      }),
    },
  }

page.onLoadFinished = function() {
  setTimeout(function() {
    page.render(env.PHANTOM_TMP_NAME)
    phantom.exit()
  }, env.PHANTOM_DELAY)
}

setTimeout(function() { phantom.exit() }, 20000)

if (env.PHANTOM_URL) {
  console.log('rendering - ', env.PHANTOM_URL)
  page.open(env.PHANTOM_URL)
} else {
  console.log('rendering - ', env.PHANTOM_HTML.substring(0, 2000))
  page.setContent(env.PHANTOM_HTML, 'http://localhost')
}


scaryshot.js
#!/usr/bin/env node

var cp = require('child_process')
var os = require('os')
var fs = require('fs')
var http = require('http')
var url = require('url')
var qs = require('querystring')
var port = process.argv[2] || 3000

var tmpdir = os.tmpdir() + '/'
var tmpid = Date.now()
var indexHtml = fs.readFileSync(__dirname + '/index.html').toString()

var mimes = {
  'pdf': 'application/pdf',
  'png': 'image/png',
  'jpeg': 'image/jpeg',
  'jpg': 'image/jpeg',
  'gif': 'image/gif',
}

function error(res, code, message) {
  res.writeHead(code, {'Content-Type': 'application/json'})
  res.end(JSON.stringify({
    status: code,
    message: message,
  }))
}

function make(opts, res) {
  if (!opts || (!opts.url && !opts.html)) {
    return error(res, 400, 'The "url" or "html" parameter is required')
  }

  var type = opts.type || 'pdf'
  if (!mimes[type]) {
    return error(res, 400, 'Type "' + type + '" is not supported.  The following types are supported: ' + Object.keys(mimes))
  }

  var name = opts.name || 'output'
  var path = tmpdir + 'phantom-' + (tmpid++) + '.' + type

  if (opts.url && !/^http/.test(opts.url)) {
    opts.url = 'http://' + opts.url
  }

  var env = JSON.stringify({
    PHANTOM_URL: opts.url || '',
    PHANTOM_TYPE: type,
    PHANTOM_TMP_NAME: path,
    PHANTOM_HTML: opts.html || '',
    PHANTOM_DELAY: opts.delay || 0,
    PHANTOM_HEADER: opts.header || '',
    PHANTOM_FOOTER: opts.footer || '',
    PHANTOM_ORIENTATION: opts.orientation || '',
  })

  var child = cp.spawn('phantomjs',
    [__dirname + '/render.js'],
    {stdio: ['pipe', process.stdout, process.stderr]}
  )

  child.stdin.write(env)
  child.stdin.end()
  child.on('error', function() {
    child.kill()
    error(res, 500, 'Unable to render webpage')
  })
  child.on('exit', function() {
    if (res.headersSent) return // sent an error

    res.writeHead(200, {
      'Content-Type': mimes[type],
      'Content-Disposition': 'attachment; filename=' + name + '.' + type,
    })
    fs.createReadStream(path)
      .on('error', function() {
        res.end()
      })
      .on('end', function() {
        fs.unlink(path, function(er) { // clean up tmp file
          if (er) return console.error(er)
          else console.log('unlinked', path)
        })
      })
      .pipe(res)
  })
}

var server = module.exports = http.createServer(function(req, res) {
  var parsed = url.parse(req.url, true)
  if (req.method == 'GET') return route(parsed.query)

  var data = ''
  req.setEncoding('utf8')
  req.on('data', function(chunk) { data += chunk })
  req.on('end', function() { route(qs.parse(data)) })
  req.on('error', console.error)

  function route(query) {
    switch (parsed.pathname) {
    case '/generate':
    case '/generate/':
      make(query, res)
      break
    default:
      res.writeHead(200, {'Content-Type': 'text/html'})
      res.end(indexHtml)
    }
  }
})

if (!module.parent) {
  server.listen(port)
  console.log('Scaryshot listening on port', port)
}
